File: frontend/src/lib/api/lms.ts
import { apiFetch } from "@/lib/api/client";
import type { ContentOut, ContentUploadOut, ContentAccessOut, LMSAssignmentOut, Page } from "@/lib/api/types";

// ============ Contents ============

export async function listContents() {
  return apiFetch<ContentOut[]>("console", "/lms/contents");
}

export async function createContent(payload: {
  title: string;
  description?: string;
  content_type: string;
  url?: string | null;
  duration_minutes?: number;
  is_platform_managed?: boolean;
}) {
  return apiFetch<ContentOut>("console", "/lms/contents", { method: "POST", body: JSON.stringify(payload) });
}

export async function updateContent(contentId: string, payload: {
  title?: string;
  description?: string;
  url?: string;
  duration_minutes?: number;
  is_active?: boolean;
}) {
  return apiFetch<ContentOut>("console", `/lms/contents/${contentId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function deleteContent(contentId: string) {
  return apiFetch<{ deleted: boolean }>("console", `/lms/contents/${contentId}`, { method: "DELETE" });
}

export async function createContentUpload(payload: {
  title: string;
  description?: string;
  filename: string;
  mime_type: string;
  duration_seconds?: number;
  is_platform_managed?: boolean;
}) {
  return apiFetch<ContentUploadOut>("console", "/lms/contents/upload", { method: "POST", body: JSON.stringify(payload) });
}

export async function getContentAccess(content_id: string) {
  return apiFetch<ContentAccessOut>("console", `/lms/contents/${content_id}/access`);
}

// ============ Learning Paths ============

export interface LearningPathItemOut {
  id: string;
  content_item_id: string;
  order_index: number;
  content_title?: string | null;
}

export interface LearningPathOut {
  id: string;
  tenant_id?: string | null;
  title: string;
  description?: string | null;
  is_platform_managed: boolean;
  is_active: boolean;
  items: LearningPathItemOut[];
  created_at: string;
}

export async function listLearningPaths() {
  return apiFetch<LearningPathOut[]>("console", "/lms/learning-paths");
}

export async function getLearningPath(pathId: string) {
  return apiFetch<LearningPathOut>("console", `/lms/learning-paths/${pathId}`);
}

export async function createLearningPath(payload: {
  title: string;
  description?: string;
  content_item_ids: string[];
}) {
  return apiFetch<LearningPathOut>("console", "/lms/learning-paths", { method: "POST", body: JSON.stringify(payload) });
}

export async function updateLearningPath(pathId: string, payload: {
  title?: string;
  description?: string;
  content_item_ids?: string[];
}) {
  return apiFetch<LearningPathOut>("console", `/lms/learning-paths/${pathId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function deleteLearningPath(pathId: string) {
  return apiFetch<{ deleted: boolean }>("console", `/lms/learning-paths/${pathId}`, { method: "DELETE" });
}

// ============ Assignments ============

export async function listAssignments(params?: { employee_id?: string; org_unit_id?: string; content_item_id?: string; status?: string; limit?: number; offset?: number }) {
  const qs = new URLSearchParams();
  if (params?.employee_id) qs.set("employee_id", params.employee_id);
  if (params?.org_unit_id) qs.set("org_unit_id", params.org_unit_id);
  if (params?.content_item_id) qs.set("content_item_id", params.content_item_id);
  if (params?.status) qs.set("status", params.status);
  qs.set("limit", String(params?.limit ?? 50));
  qs.set("offset", String(params?.offset ?? 0));
  return apiFetch<Page<LMSAssignmentOut>>("console", `/lms/assignments?${qs.toString()}`);
}

export async function createAssignment(payload: { content_item_id?: string; learning_path_id?: string; employee_id?: string | null; org_unit_id?: string | null }) {
  return apiFetch<{ id: string }>("console", "/lms/assignments", { method: "POST", body: JSON.stringify(payload) });
}

export async function updateAssignment(assignmentId: string, payload: { due_at?: string; status?: string }) {
  return apiFetch<LMSAssignmentOut>("console", `/lms/assignments/${assignmentId}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export async function deleteAssignment(assignmentId: string) {
  return apiFetch<{ deleted: boolean }>("console", `/lms/assignments/${assignmentId}`, { method: "DELETE" });
}

export async function bulkCreateAssignments(payload: {
  content_item_id?: string;
  learning_path_id?: string;
  employee_ids?: string[];
  org_unit_ids?: string[];
}) {
  return apiFetch<{ created: number }>("console", "/lms/assignments/bulk", { method: "POST", body: JSON.stringify(payload) });
}

// ============ Completions ============

export async function createCompletion(payload: { assignment_id: string; completion_method?: string }) {
  return apiFetch<{ id: string }>("console", "/lms/completions", { method: "POST", body: JSON.stringify(payload) });
}

// ============ Stats ============

export interface LMSStatsOut {
  total_contents: number;
  total_assignments: number;
  total_completed: number;
  total_in_progress: number;
  completion_rate: number;
  contents_by_type: Record<string, number>;
  assignments_by_status: Record<string, number>;
}

export async function getLMSStats() {
  return apiFetch<LMSStatsOut>("console", "/lms/stats");
}

File: backend/app/schemas/lms.py
from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, Field
from uuid import UUID
from typing import Optional


class ContentCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)
    content_type: str = Field(default="video", description="video|pdf|link")
    url: Optional[str] = Field(default=None, description="Obrigatório para content_type=link. Para uploads, deixe vazio.")
    duration_minutes: Optional[int] = None
    is_platform_managed: bool = False


class ContentUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=2, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)
    url: Optional[str] = None
    duration_minutes: Optional[int] = None
    is_active: Optional[bool] = None


class ContentUploadCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)
    filename: str = Field(..., min_length=1, max_length=255)
    mime_type: str = Field(default="application/octet-stream", max_length=120)
    duration_seconds: Optional[int] = None
    is_platform_managed: bool = False


class ContentUploadOut(BaseModel):
    content_id: UUID
    upload_url: str
    method: str = "PUT"
    expires_in_seconds: int


class ContentAccessOut(BaseModel):
    content_id: UUID
    access_url: str
    expires_in_seconds: int


class ContentOut(BaseModel):
    id: UUID
    title: str
    description: Optional[str] = None
    content_type: str
    url: Optional[str] = None
    storage_key: Optional[str] = None
    duration_minutes: Optional[int] = None
    is_platform_managed: bool
    is_active: bool


class LearningPathCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)
    content_item_ids: list[UUID] = Field(default_factory=list, description="Ordered list of content IDs")


class LearningPathUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=2, max_length=200)
    description: Optional[str] = Field(default=None, max_length=500)
    content_item_ids: Optional[list[UUID]] = None


class LearningPathItemOut(BaseModel):
    id: UUID
    content_item_id: UUID
    order_index: int
    content_title: Optional[str] = None


class LearningPathOut(BaseModel):
    id: UUID
    tenant_id: Optional[UUID] = None
    title: str
    description: Optional[str] = None
    is_platform_managed: bool
    is_active: bool = True
    items: list[LearningPathItemOut] = Field(default_factory=list)
    created_at: datetime


class AssignmentCreate(BaseModel):
    content_item_id: Optional[UUID] = None
    learning_path_id: Optional[UUID] = None
    employee_id: Optional[UUID] = None
    org_unit_id: Optional[UUID] = None


class AssignmentUpdate(BaseModel):
    due_at: Optional[datetime] = None
    status: Optional[str] = None


class AssignmentOut(BaseModel):
    id: UUID
    content_item_id: Optional[UUID] = None
    learning_path_id: Optional[UUID] = None
    employee_id: Optional[UUID] = None
    org_unit_id: Optional[UUID] = None
    due_at: Optional[datetime] = None
    status: str
    created_at: datetime

    progress_seconds: Optional[int] = None
    duration_seconds: Optional[int] = None
    completed_at: Optional[datetime] = None


class BulkAssignmentCreate(BaseModel):
    content_item_id: Optional[UUID] = None
    learning_path_id: Optional[UUID] = None
    employee_ids: Optional[list[UUID]] = Field(default=None)
    org_unit_ids: Optional[list[UUID]] = Field(default=None)


class CompletionCreate(BaseModel):
    assignment_id: UUID
    completion_method: str = "manual"


class ProgressUpdate(BaseModel):
    assignment_id: UUID
    position_seconds: int = Field(..., ge=0)
    duration_seconds: Optional[int] = Field(default=None, ge=0)


class ProgressOut(BaseModel):
    assignment_id: UUID
    employee_id: UUID
    position_seconds: int
    duration_seconds: Optional[int] = None
    last_event_at: datetime


class LMSStatsOut(BaseModel):
    total_contents: int = 0
    total_assignments: int = 0
    total_completed: int = 0
    total_in_progress: int = 0
    completion_rate: float = 0.0
    contents_by_type: dict[str, int] = Field(default_factory=dict)
    assignments_by_status: dict[str, int] = Field(default_factory=dict)
